pkg/agent/controller/serviceimport.go

```diff
@@ -0,0 +1,224 @@
+				c.queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			klog.Infof("ServiceImport %q deleted", key)
+			if err == nil {
+				var si *lighthousev2a1.ServiceImport
+				var ok bool
+				if si, ok = obj.(*lighthousev2a1.ServiceImport); !ok {
+					tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+					if !ok {
+						klog.Errorf("Failed to get deleted serviceimport object for key %s, serviceImport %v", key, si)
+						return
+					}
+
+					si, ok = tombstone.Obj.(*lighthousev2a1.ServiceImport)
+
+					if !ok {
+						klog.Errorf("Failed to convert deleted tombstone object %v  to serviceimport", tombstone.Obj)
+						return
+					}
+				}
+				if si.Spec.Type != lighthousev2a1.Headless {
+					return
+				}
+				c.endpointControllerDeleted.Store(si, key)
+				c.queue.AddRateLimited(key)
+			}
+		},
+	})
+
+	go c.serviceInformer.Run(stopCh)
+	go c.runServiceImportWorker()
+
+	go func(stopCh <-chan struct{}) {
+		<-stopCh
+		c.queue.ShutDown()
+
+		klog.Infof("ServiceImport Controller stopped")
+	}(stopCh)
+
+	return nil
+}
+
+func (c *ServiceImportController) runServiceImportWorker() {
+	for {
+		keyObj, shutdown := c.queue.Get()
+		if shutdown {
+			klog.Infof("Lighthouse watcher for ServiceImports stopped")
+			return
+		}
+
+		key := keyObj.(string)
+
+		func() {
+			defer c.queue.Done(key)
+			obj, exists, err := c.serviceInformer.GetIndexer().GetByKey(key)
+
+			if err != nil {
+				klog.Errorf("Error retrieving the object with store is  %v from the cache: %v", c.serviceInformer.GetIndexer().ListKeys(), err)
+				// requeue the item to work on later
+				c.queue.AddRateLimited(key)
+
+				return
+			}
+
+			c.queue.Forget(key)
+
+			if exists {
+				c.serviceImportCreatedOrUpdated(obj, key)
+			} else {
+				c.serviceImportDeleted(key)
+			}
+		}()
+	}
+}
+
+func (c *ServiceImportController) serviceImportCreatedOrUpdated(obj interface{}, key string) {
+	if _, found := c.endpointControllerCreated.Load(key); found {
+		klog.V(2).Infof("The endpointController is already running for  key %q", key)
+		return
+	}
+
+	serviceImportCreated := obj.(*lighthousev2a1.ServiceImport)
+	if serviceImportCreated.Spec.Type != lighthousev2a1.Headless {
+		return
+	}
+
+	annotations := serviceImportCreated.ObjectMeta.Annotations
+	serviceNameSpace := annotations[originNamespace]
+	serviceName := annotations[originName]
+	var service *corev1.Service
+	var isFound bool
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		var err error
+		service, err = c.kubeClientSet.CoreV1().Services(serviceNameSpace).Get(serviceName, metav1.GetOptions{})
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				isFound = false
+			}
+			return err
+		}
+		isFound = true
+		return nil
+	})
+
+	if retryErr != nil || !isFound {
+		klog.Errorf("Error retrieving the service  %q from the namespace %q : %v", serviceName, serviceNameSpace, retryErr)
+		return
+	}
```

Comment:
If the `Service` doesn't exist I guess we can ignore it since we should get a `ServiceImport` delete at some point. Otherwise we should log the error and re-queue.

Reply:
Done.